sockeye/utils.py

```diff
@@ -280,22 +280,26 @@ def topk(scores: mx.nd.NDArray,
          use_mxnet_topk: bool) -> Tuple[mx.nd.NDArray, mx.nd.NDArray, mx.nd.NDArray]:
     """
     Get the lowest k elements per sentence from a `scores` matrix.
+    At the first timestep, the shape of scores is (batch, target_vocabulary_size).
+    At subsequent steps, the shape is (batch * k, target_vocabulary_size).
 
     :param scores: Vocabulary scores for the next beam step. (batch_size * beam_size, target_vocabulary_size)
     :param k: The number of smallest scores to return.
     :param offset: Array to add to the hypothesis indices for offsetting in batch decoding.
     :param use_mxnet_topk: True to use the mxnet implementation or False to use the numpy one.
     :return: The row indices, column indices and values of the k smallest items in matrix.
     """
+    # Compute the batch size from the offsets. We don't know the batch size because it is
+    # either 1 (timestep 1) or k (timestep 2+).
     # (batch_size, beam_size * target_vocab_size)
-    folded_scores = scores.reshape((-1, k * scores.shape[-1]))
-    batch_size = folded_scores.shape[0]
+    batch_size = int(offset[-1].asscalar() / k) + 1
+    folded_scores = scores.reshape((batch_size, -1))
 
```

Comment:
I wonder if we can avoid the `asscalar()` conversion here, since we know that offset is always batch*beam:
```python
beam_size = int(offset.shape[0] / k)
scores = scores.reshape(-1, beam_size, 0) # maybe we need reshape(-4, -1, beam_size, 0). resulting shape: (batch, beam, target_vocab_size)
scores = scores.reshape(0, -3) # shape: (batch, beam*target_vocab_size)
```

Reply:
Oh, sorry, missed this one. I’ll take a look later.

Reply:
Actually, I don't understand this. Why do we need to avoid `asscalar()`? Also, your `beam_size` variable above is actually the batch size.

Reply:
Okay, I think I fixed this after swapping it back in.

Reply:
Thanks! You are right, I was confusing batch and beam here. The reason, I would like to avoid `asscalars` is the same as for `asnumpys`: they cause a synchronization between mxnets backend and python/numpy frontend and can potentially make things slower. The shape information of ndarrays is available instantaneously. Probably not too important here, but why take the 'risk'? :

Reply:
Yeah, even after noting that typo of yours, I still managed to fix what you wanted :)